Foundation/Array/Unboxed.hs

```diff
@@ -1252,3 +1254,71 @@ toHexadecimal ba
                 unsafeWrite ma dIdx     (W8# wHi)
                 unsafeWrite ma (dIdx+1) (W8# wLo)
                 loop (dIdx + 2) (sIdx+1)
+
+toBase64Internal :: PrimType ty => Addr# -> UArray ty -> Bool -> UArray Word8
+toBase64Internal table src padded
+    | len == CountOf 0 = empty
+    | otherwise = runST $ do
+        ma <- new dstLen
+        unsafeIndexer b8 (go ma)
+        unsafeFreeze ma
+  where
+    b8 = unsafeRecast src
+    !len = length b8
+    !dstLen = outputLengthBase64 padded len
+    !endOfs = Offset 0 `offsetPlusE` len
+
+    go :: MUArray Word8 s -> (Offset Word8 -> Word8) -> ST s ()
+    go !ma !getAt = loop 0 0
+      where
+        eqChar = 0x3d :: Word8
+
+        loop !sIdx !dIdx
+            | sIdx >= endOfs = return ()
+            | otherwise = do
```

Comment:
I think here, I would put the case where you going to write padding (1 & 2 bytes left) too, and that would leave the `otherwise` (i.e. at least 3 bytes) case clearer and the only case that would call `loop` again.

```
loop
    | sIdx == endOfs = return ()
    | sIdx + 3 > endOfs = ... -- no 3 bytes, padding cases
    | otherwise          = ... -- normal cases, + looping
```